Samples/WinMLSamplesGallery/WinMLSamplesGallery/Samples/Batching/Batching.xaml.cs

```diff
@@ -0,0 +1,240 @@
+using System;
+using System.Collections.Generic;
+using Microsoft.UI.Xaml;
+using Microsoft.UI.Xaml.Controls;
+using Windows.Storage;
+using Windows.Graphics.Imaging;
+using Windows.Media;
+using Microsoft.AI.MachineLearning;
+
+namespace WinMLSamplesGallery.Samples
+{
+    public sealed class EvalResult
+    {
+        public string nonBatchedAvgTime { get; set; }
+        public string batchedAvgTime { get; set; }
+        public string timeRatio { get; set; }
+    }
+
+    public sealed partial class Batching : Page
+    {
+        private LearningModelSession nonBatchingSession_;
+        private LearningModelSession BatchingSession_;
+        float avgNonBatchedDuration = 0;
+        float avgBatchDuration = 0;
+        static bool navigatingAwayFromPage = false;
```

Comment:
I think there's a race on `navigatingAwayFromPage` and some synchronization missing, since it's [being set in another task](https://github.com/microsoft/Windows-Machine-Learning/pull/413/files#diff-f3358a7e32e2efb662773a0cb7f6b8f4fe0e52ff9c75f5e9c89861949f99adfaR37-R42), and you expect it to be updated in this task.

Take a look at [Interlocked](https://docs.microsoft.com/en-us/dotnet/api/system.threading.interlocked?view=net-5.0) or mutexes in C#. Also, it's a good idea to abstract synchronization in another method instead of having `if (navigatingAwayFromPage)` in multiple places. Race conditions are nasty and complicated and you don't want multithreaded code spread out unnecessarily.

Reply:
Also, there's classes in C# dedicated to Task cancellation: https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/task-cancellation . It's preferrable to use them if possible :)